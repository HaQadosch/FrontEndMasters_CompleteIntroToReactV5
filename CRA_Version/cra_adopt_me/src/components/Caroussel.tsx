import React, { useState, ReactEventHandler, useEffect } from 'react';
import { Photo } from '@frontendmasters/pet';

export const Caroussel: React.FC<{ photos: Photo[] }> = ({ photos: media }) => {
  const smallDefault = 'http://placecorgi.com/300/300';
  const largeDefault = 'http://placecorgi.com/600/600';
  const [photos, setPhotos] = useState<Photo[]>(media);
  const [active, setActive] = useState<number>(0);

  useEffect(() => {
    setPhotos(media.map(photo => ({ large: largeDefault, small: smallDefault, ...photo })));
  }, [media]);

  const onImgClick = (index: number): ReactEventHandler => ({ target }: React.SyntheticEvent) => {
    setActive(index);
  };

  return (
    <div className='carousel'>
      <img src={photos[active].large} alt='active pet' />
      <div className='carousel-smaller'>
        {photos.map((photo, index) => (
          <img
            key={index}
            src={photo.small}
            alt='small pet'
            className={index === active ? 'active' : ''}
            onClick={onImgClick(index)}
          />
        ))}
      </div>
    </div>
  );
};
