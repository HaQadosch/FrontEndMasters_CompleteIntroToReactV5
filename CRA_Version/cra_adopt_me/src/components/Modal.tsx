import React, { useRef, useEffect } from 'react';
import { createPortal } from 'react-dom';

export const Modal: React.FC = ({ children }) => {
  const elt = useRef<HTMLDivElement>();
  if (!elt.current) {
    const div = document.createElement('div');
    elt.current = div;
  }

  useEffect(() => {
    const modalRoot = document.querySelector('#modal');
    if (modalRoot && elt.current) {
      modalRoot.appendChild(elt.current);
    }
    return () => {
      if (modalRoot && elt.current) {
        modalRoot.removeChild(elt.current);
      }
    };
  }, []);

  return createPortal(<div>{children}</div>, elt.current);
};
