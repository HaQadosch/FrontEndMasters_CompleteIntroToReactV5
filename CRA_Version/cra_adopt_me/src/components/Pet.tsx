import React from 'react';
import { Animal } from '@frontendmasters/pet';
import { Link } from '@reach/router';

export const Pet: React.FC<Animal> = ({
  id,
  name,
  type: animal,
  breeds: { primary: breed } = { primary: '' },
  photos: media = [],
  contact: { address } = { address: { city: '', state: '' } },
}) => {
  let hero = media.length ? media[0].small : 'http://placecorgi.com/300/300';
  return (
    <Link to={`/details/${id}`} className='pet'>
      <div className='image-container'>
        <img src={hero} alt={name} />
      </div>
      <div className='info'>
        <h1>{name}</h1>
        <h2>{`${animal} - ${breed} - ${address.city}, ${address.state}`}</h2>
      </div>
    </Link>
  );
};
