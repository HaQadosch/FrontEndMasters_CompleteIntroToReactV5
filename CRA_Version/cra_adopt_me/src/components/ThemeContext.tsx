import React, { createContext, SetStateAction, Dispatch } from 'react';

// export const ThemeContext: React.Context<(string | (() => string))[]> = createContext(['green', () => ''])
export const ThemeContext: React.Context<(string | (Dispatch<SetStateAction<string>>))[]> = createContext([
  'green',
  (_: SetStateAction<string>) => {},
]);
