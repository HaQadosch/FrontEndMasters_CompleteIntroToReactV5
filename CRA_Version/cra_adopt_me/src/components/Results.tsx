import React from 'react';
import { Pet } from './Pet';
import { Animal } from '@frontendmasters/pet';

export const Results: React.FC<{ pets: Animal[] }> = ({ pets }) => {
  return (
    <div className='search'>
      {pets.length === 0 ? <h1>No pets found.</h1> : pets.map(({ id, ...pet }) => <Pet key={id} id={id} {...pet} />)}
    </div>
  );
};
