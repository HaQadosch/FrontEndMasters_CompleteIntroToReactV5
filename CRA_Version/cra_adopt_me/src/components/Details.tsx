import React, { useEffect, useState, useReducer, Dispatch, useContext } from 'react';
import pet, { Animal, Address, Breeds } from '@frontendmasters/pet';
import { Caroussel } from './Caroussel';
import { ThemeContext } from './ThemeContext';
import { navigate } from '@reach/router';
import { Modal } from './Modal';

type DetailsAnimal = Pick<Animal, 'name' | 'type' | 'description' | 'photos' | 'url'> &
  Pick<Address, 'city' | 'state'> &
  Pick<Breeds, 'primary'>;

const animalReducer = (currentState: DetailsAnimal, newState: Animal): DetailsAnimal => {
  const {
    contact: {
      address: { city, state },
    },
    breeds: { primary },
    ...restState
  } = newState;
  return { ...currentState, ...restState, city, state, primary };
};

export const Details: React.FC<{ path: string; id?: string }> = ({ id = '' }) => {
  const [loading, setLoading] = useState<boolean>(true);
  const [, /*error*/ setError] = useState<Error>();
  const [theme /*, setTheme*/] = useContext<(string | any)[]>(ThemeContext);

  const [showModal, setShowModal] = useState<boolean>(false);
  const toggleModal = () => {
    setShowModal(!showModal);
  };

  const [{ name, type, city, state, description, photos, primary: breed, url }, setState]: [
    DetailsAnimal,
    Dispatch<Animal>
  ] = useReducer(animalReducer, {
    url: '',
    name: '',
    type: '',
    description: '',
    photos: [],
    city: '',
    state: '',
    primary: '',
  });

  const adopt = () => navigate(url);

  useEffect(() => {
    pet
      .animal(parseInt(id))
      .then(({ animal }) => {
        setState(animal);
        setLoading(false);
      })
      .catch(petErr => setError(petErr));
  }, [id]);

  return loading ? (
    <h1>
      <span role='img' aria-label='sand-glass'>
        âŒ›
      </span>
      Loading...
    </h1>
  ) : (
    <div className='details'>
      <Caroussel photos={photos} />
      <div>
        <h1>{name}</h1>
        <h2>{`${type} - ${breed} - ${city}, ${state}`}</h2>
        <button onClick={toggleModal} style={{ backgroundColor: theme }}>{`Adopt ${name}`}</button>
        <p>{description}</p>
        {showModal ? (
          <Modal>
            <div>
              <h1>Would you like to adopt {name} ?</h1>
              <div className='buttons'>
                <button onClick={adopt}>
                  Yes{' '}
                  <span role='img' aria-label='shiny heart'>
                    ðŸ’–
                  </span>
                </button>
                <button onClick={toggleModal}>
                  No{' '}
                  <span role='img' aria-label='broken heart'>
                    ðŸ’”
                  </span>
                </button>
              </div>
            </div>
          </Modal>
        ) : null}
      </div>
    </div>
  );
};
