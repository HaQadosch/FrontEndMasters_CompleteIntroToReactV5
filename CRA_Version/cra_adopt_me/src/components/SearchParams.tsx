import React, { useState, useEffect, useContext } from 'react';
import pet, { ANIMALS, AnimalsResponse, Animal, AnimalsParams } from '@frontendmasters/pet';
import { Results } from './Results';
import { useDropdown } from '../utils/useDropdown';
import { ThemeContext } from './ThemeContext';

export const SearchParams: React.FC<{ path: string }> = () => {
  const [location, setLocation] = useState<string>('Seattle, WA');
  const [breeds, setBreeds] = useState<string[]>([]);
  const [animal /*setAnimal*/, , AnimalDropdown] = useDropdown('Animal', 'all', ANIMALS);
  const [breed, setBreed, BreedDropdown] = useDropdown('Breed', 'dog', breeds);
  const [pets, setPets] = useState<Animal[]>([]);
  const [theme, setTheme] = useContext<(string | any)[]>(ThemeContext);

  const requestPet = async (requestParams: AnimalsParams) => {
    const { animals }: AnimalsResponse = await pet.animals(requestParams);
    setPets(animals || []);
  };

  useEffect(() => {
    setBreeds([]);
    setBreed('');

    pet.breeds(animal).then(({ breeds: petBreeds }) => {
      if (petBreeds) {
        const breedsNames = petBreeds.map(({ name }) => name);
        setBreeds(breedsNames);
      }
    }, console.error);
  }, [animal, setBreed]);

  const onInputLocationChange: React.ChangeEventHandler<HTMLInputElement> = ({ target: { value } }) =>
    setLocation(value);

  const onFormSubmit: React.ChangeEventHandler<HTMLFormElement> = evt => {
    evt.preventDefault();
    requestPet({
      location,
      breed,
      type: animal,
    });
  };

  const onThemeSelectChange: React.ChangeEventHandler<HTMLSelectElement> = ({ target: { value } }) => {
    setTheme(value);
  };

  return (
    <div className='search-params'>
      <form action='/' onSubmit={onFormSubmit}>
        <label htmlFor='location'>
          Location
          <input
            type='text'
            name='location'
            id='location'
            value={location}
            onChange={onInputLocationChange}
            placeholder='location'
          />
        </label>
        <AnimalDropdown />
        <BreedDropdown />
        <label htmlFor='theme'>
          Theme
          <select name='theme' id='theme' value={theme} onChange={onThemeSelectChange} onBlur={onThemeSelectChange}>
            <option value='pere'>peru</option>
            <option value='#bada55'>bada55</option>
            <option value='mediumorchid'>medium orchid</option>
          </select>
        </label>
        <button type='submit' style={{ backgroundColor: theme }}>
          Submit
        </button>
      </form>
      <Results pets={pets} />
    </div>
  );
};
