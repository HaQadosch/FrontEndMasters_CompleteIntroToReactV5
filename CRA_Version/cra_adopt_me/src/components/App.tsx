import React, { useState } from 'react';
import { Router, Link } from '@reach/router';
import './App.css';
import { SearchParams } from './SearchParams';
import { Details } from './Details';
import { ThemeContext } from './ThemeContext';

export const App: React.FC = () => {
  const [color, setColor] = useState<string>('darkblue');

  return (
    <ThemeContext.Provider value={[color, setColor]}>
      <div>
        <header>
          <Link to={'/'}>Adopt me!</Link>
        </header>
        <React.StrictMode>
          <Router>
            <SearchParams path='/' />
            <Details path='/details/:id' id='' />
          </Router>
        </React.StrictMode>
      </div>
    </ThemeContext.Provider>
  );
};
