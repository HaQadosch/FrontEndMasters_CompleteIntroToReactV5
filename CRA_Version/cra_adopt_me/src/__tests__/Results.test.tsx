import React from 'react';
import ReactDOM from 'react-dom';
import { render } from 'react-testing-library';
import { Results } from '../components/Results';
import { Animal } from '@frontendmasters/pet';
import 'jest';

const testPet: Animal = {
  id: 44604528,
  organization_id: 'WA41',
  url:
    'https://www.petfinder.com/bird/russell-crow-44604528/wa/silverdale/kitsap-humane-society-wa41/?referrer_id=44e48166-8259-4b23-8420-bbd2dc9c2079',
  type: 'Bird',
  species: 'Chicken',
  breeds: {
    primary: 'Chicken',
    secondary: null,
    mixed: true,
    unknown: false,
  },
  colors: {
    primary: null,
    secondary: null,
    tertiary: null,
  },
  age: 'Adult',
  gender: 'Male',
  size: 'Medium',
  coat: null,
  attributes: {
    spayed_neutered: false,
    house_trained: false,
    declawed: null,
    special_needs: false,
    shots_current: true,
  },
  environment: {
    children: null,
    dogs: null,
    cats: null,
  },
  tags: [],
  name: 'Russell Crow',
  description:
    'Meet Russell Crow! He is a gorgeous adult rooster that could be a great addition to your flock or farm....',
  photos: [
    {
      small: 'https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/44604528/1/?bust=1556978254&width=100',
      medium: 'https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/44604528/1/?bust=1556978254&width=300',
      large: 'https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/44604528/1/?bust=1556978254&width=600',
      full: 'https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/44604528/1/?bust=1556978254',
    },
    {
      small: 'https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/44604528/2/?bust=1556978268&width=100',
      medium: 'https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/44604528/2/?bust=1556978268&width=300',
      large: 'https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/44604528/2/?bust=1556978268&width=600',
      full: 'https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/44604528/2/?bust=1556978268',
    },
  ],
  status: 'adoptable',
  published_at: '2019-05-03T12:26:26+0000',
  contact: {
    email: 'CustomerService@kitsap-humane.org',
    phone: '(360) 692-6977',
    address: {
      address1: '9167 Dickey Road NW',
      address2: null,
      city: 'Silverdale',
      state: 'WA',
      postcode: '98383',
      country: 'US',
    },
  },
  _links: {
    self: {
      href: '/v2/animals/44604528',
    },
    type: {
      href: '/v2/types/bird',
    },
    organization: {
      href: '/v2/organizations/wa41',
    },
  },
};

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Results pets={[testPet]} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('Pet matches the snapshot', () => {
  const { debug, asFragment } = render(<Results pets={[testPet]} />);
  expect(asFragment()).toMatchSnapshot();
  // debug(asFragment());
});
