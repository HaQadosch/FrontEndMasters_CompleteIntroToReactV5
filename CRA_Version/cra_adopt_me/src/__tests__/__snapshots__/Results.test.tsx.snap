// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pet matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="search"
  >
    <a
      class="pet"
      href="/details/44604528"
    >
      <div
        class="image-container"
      >
        <img
          alt="Russell Crow"
          src="https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/44604528/1/?bust=1556978254&width=100"
        />
      </div>
      <div
        class="info"
      >
        <h1>
          Russell Crow
        </h1>
        <h2>
          Bird - Chicken - Silverdale, WA
        </h2>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
